packages/nextjs/src/config/wrappers/utils/responseEnd.ts

```diff
@@ -0,0 +1,70 @@
+import { flush } from '@sentry/node';
+import { Transaction } from '@sentry/types';
+import { fill, logger } from '@sentry/utils';
+
+import { AugmentedResponse, ResponseEndMethod, WrappedResponseEndMethod } from '../types';
+
+/**
+ * Wrap `res.end()` so that it closes the transaction and flushes events before letting the request finish.
+ *
+ * Note: This wraps a sync method with an async method. While in general that's not a great idea in terms of keeping
+ * things in the right order, in this case it's safe, because the native `.end()` actually *is* (effectively) async, and
+ * its run actually *is* (literally) awaited, just manually so (which reflects the fact that the core of the
+ * request/response code in Node by far predates the introduction of `async`/`await`). When `.end()` is done, it emits
+ * the `prefinish` event, and only once that fires does request processing continue. See
+ * https://github.com/nodejs/node/commit/7c9b607048f13741173d397795bac37707405ba7.
+ *
+ * Also note: `res.end()` isn't called until *after* all response data and headers have been sent, so blocking inside of
+ * `end` doesn't delay data getting to the end user. See
+ * https://nodejs.org/api/http.html#responseenddata-encoding-callback.
+ *
+ * @param transaction The transaction tracing request handling
+ * @param res: The request's corresponding response
+ */
+export function autoEndTransactionOnResponseEnd(transaction: Transaction, res: AugmentedResponse): void {
+  res.__sentryTransaction = transaction;
+
+  const wrapEndMethod = (origEnd: ResponseEndMethod): WrappedResponseEndMethod => {
+    return async function sentryWrappedEnd(this: AugmentedResponse, ...args: unknown[]) {
+      await finishTransaction(this);
+      await flushQueue();
+
+      return origEnd.call(this, ...args);
+    };
+  };
+
+  // Prevent double-wrapping
+  if (!(res.end as WrappedResponseEndMethod).__sentry_original__) {
+    fill(res, 'end', wrapEndMethod);
+  }
+}
+
+/** Finish the given response's transaction and set HTTP status data */
+export async function finishTransaction(res: AugmentedResponse): Promise<void> {
+  const { __sentryTransaction: transaction } = res;
+
+  if (transaction) {
+    transaction.setHttpStatus(res.statusCode);
+
+    // Push `transaction.finish` to the next event loop so open spans have a better chance of finishing before the
+    // transaction closes, and make sure to wait until that's done before flushing events
+    const transactionFinished: Promise<void> = new Promise(resolve => {
+      setImmediate(() => {
+        transaction.finish();
+        resolve();
+      });
+    });
```

Comment:
Do we actually know whether this is necessary? In which scenarios would spans be finished *after* `res.end()` has beed called?

Reply:
To be fair, in nextjs I'm not sure it is. It's logic originally from https://github.com/getsentry/sentry-javascript/blob/a5bfa8013ab84c6e43a7d48b9a0da197d2b32443/packages/node/src/handlers.ts#L86-L94, which got copied over into the original version of `withSentry`.

UPDATE: I just checked the blame, and it turns out that for express, the `setImmediate()` was introduced in https://github.com/getsentry/sentry-javascript/pull/3022 to fix https://github.com/getsentry/sentry-javascript/issues/3001. I don't see any analogs to that problem for nextjs, so yeah, I think we can get rid of it. Will do.

Reply:
Awesome thank you!

Reply:
On second thought, what if the user has manually added a span which finishes on `res.end()`? Or what if an integration does? As long as we drop spans which haven't finished, part of me feels like it can't hurt to leave this in. WDYT?

Reply:
For some reason I am just weirded out by this. It *shouldn't* be necessary. I am a little bit scared that this is the reason serverside tracing wasn't working in the first place. I would like to test this properly before we include it.

As for

> what if the user has manually added a span which finishes on res.end()? Or what if an integration does?

I think these are pretty niche cases.

Reply:
> I am a little bit scared that this is the reason serverside tracing wasn't working in the first place.

I still don't have good theories on that (among other things, why it works reliably when using the debugger, but not always when not), but in that case the transaction wasn't even starting - instead the code was falling off somewhere in `instrumentServer`. So I can't see how this would have had any effect there.

> It shouldn't be necessary.
> I think these are pretty niche cases. (The integration one not quite but it's something we generally own ourselves)

Agreed that right now there's nothing that I can see in the nextjs SDK which makes it necessary. And I just looked at all of the node integrations and I _think_ none of them needs it*. But I don't necessarily want to count on the fact that every future person who writes a node integration, even if they're on our team, will remember to either a) stay away from ending a span on response end or b) remember to put the `setImmediate` back in if they do.

\*The only exception is the express integration, which someone could conceivably use if they're using a custom server. Then again, we've made no effort to support that use case, so we might not care that much. 

> For some reason I am just weirded out by this.

Can you say more about why? It's been running like this successfully in `withSentry` for the last 18 months, so I'm not sure I understand the bad feeling you're getting.

Reply:
Ok let's leave it. I admit my sketched-out-ness is a bit irrational and just a gut-feeling. Still kinda sceptical but ok with it.